### PR TITLE
Introduce more specific values for RunMode and rename it to ComponentUseCase

### PR DESCRIPTION
**Pull request by gadomsky**

### CODE REVIEW

File: api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala [hunk 1]
```diff
@@ -3,16 +3,21 @@ package pl.touk.nussknacker.engine.api.process
...
+  * In this mode real implementation of node should be used.</li>
   * <li>Test - process is run in test mode to collect results for test data. Can be used to stub real implementation.</li>
   * </ul>
   */
 sealed trait RunMode
 
 object RunMode {
 
-  case object Normal extends RunMode
+  sealed trait Normal extends RunMode
```
Comment (arkadius) [inline api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala:15]: Do u see some advantage in grouping four modes below to "normal" mode other than for backward compatibility? IMO it is not straightforward why e.g. TestDataGeneration is normal mode and test is test mode

File: api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala [hunk 2]
```diff
@@ -3,16 +3,21 @@ package pl.touk.nussknacker.engine.api.process
...
+
+  case object Engine extends Normal
+  case object Validation extends Normal
```
Comment (arkadius) [inline api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala:18]: Validation doesn't cause that node will be "run". What case do you see that user will need to handle it in component?

File: api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala [hunk 3]
```diff
@@ -3,16 +3,21 @@ package pl.touk.nussknacker.engine.api.process
...
+  case object ServiceQuery extends Normal
+  case object TestDataGeneration extends Normal
```
Comment (arkadius) [inline api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala:20]: The same as above . I think that exposing too many modes will cause that consumer of those modes will be confused how they should be handled. Also when we change implementation details, some of those modes will go away and we will be hard to migrate to new approach. Maybe it should be extracted only one new mode ServiceQuery which indeed is in context of running component and validation / testdatageneration should be hidden as a normal mode as a implementation detail? (also removed from documentation)

File: api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala (hunk 2 above)
Comment (gadomsky) [reply api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala:18]: Actually I see that from diffrent angle - I care only about `Engine` mode, to be able to distinguish that my code is run via engine vs designer. 
What mode do you think should I use while performing validation?

File: api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala (hunk 1 above)
Comment (gadomsky) [reply api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala:15]: No, it was only for sake of backwards compatibility.

File: docs/MigrationGuide.md [hunk 4]
```diff
@@ -9,7 +9,8 @@ To see the biggest differences please consult the [changelog](Changelog.md).
   * Replace scenario jsonString by GraphProcess at DeploymentManager, ProcessConfigEnricherInputData
 * [#2773](https://github.com/TouK/nussknacker/pull/2773) Using VersionId / ProcessId / ProcessName instead of Long or String:
   * `PullProcessRepository` API was changed, right now we use VersionId instead of Long
-  
+* [#2830](https://github.com/TouK/nussknacker/pull/2830) `RunMode.Normal` becomes sealed trait and has specific values: Engine, Validation, ServiceQuery, TestDataGeneration
```
Comment (dswiecki) [inline docs/MigrationGuide.md:12]: copy this to changelog

File: api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala (hunk 2 above)
Comment (arkadius) [reply api/src/main/scala/pl/touk/nussknacker/engine/api/process/RunMode.scala:18]: We agree that the real usage is not only for run part but also for definition part and runMode is passed that way to both places. So the only problem was Run in name of class. Let's rename it to ComponentUsage or similar

**Review by arkadius — APPROVED** [review]

File: docs/MigrationGuide.md (hunk 4 above)
Comment (gadomsky) [reply docs/MigrationGuide.md]: done

**Pull request merged**